<a href="https://colab.research.google.com/github/AdrianJimenezNieto/ml-especializacion/blob/main/2_4_arboles_decision.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Actividad 2.4 - Arboles de Desision / Bosques Aleatorios

  - En esta practica se desarrollaran dos tecnicas de aprendizaje automatico, los arboles de decision y los bosques aleatorios.
  - Dichos modelos son modelos de clasificacion que aplicaremos al conjunto de datos ejemplo que trae Scikit Learn de Iris.

  ## Objetivos
    1. Implementar y ajustar los modelos.
    2. Visualizar y comprender la estructura de los arboles de decision para interpretar el proceso de clasificacion.
    3. Evaluar el rendimiento de ambos modelos mediante metricas como la matriz de confusion y el informe de clasificacion.
    4. Comprender y calcular la importancia de las caracteristicas para cada modelo para identificar cuales son mas influyentes.
    5. Comparar el desempeño de los árboles de decisón y los bosques aleatorios en términos de precisión, capacidad de generalización y capacidad para manejar la importancia de características.



# **ÁRBOLES DE DECISIÓN**

### **Importamos las librerías y el set de datos (dataset iris).**

In [2]:
# Importamos las librerías
import pandas as pd
import numpy as np
import sklearn as sk

In [4]:
# Nos traemos el dataset de iris de la librería sklearn
iris = sk.datasets.load_iris()

# Lo convertimos a un DataFrame de pandas para trabajarlo
df = pd.DataFrame(
    data= iris.data,
    columns= iris.feature_names
)

# Añadimos la columna con las etiquetas (clases)
df["target"] = iris.target

# Añadimos los nombres de las especies en vez de solo números para claridad
df["species"] = df["target"].apply(lambda x: iris.target_names[x])

# Visualizamos las 5 primeras filas del df
df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   target species  
0       0  setosa  
1       0  setosa  
2       0  setosa  
3       0  setosa  
4       0  setosa

### **División del conjunto de datos en entrenamiento y prueba.**

- Antes de lanzar el `train_test_split` a lo loco, vamos a mirar el tamaño de los datos de estos y, en función de estos decidimos qué porcentaje razonable usar para entrenamiento y pruebas.




In [6]:
# Ver el tamaño del dataset
print("Número de filas:", df.shape[0])
print("Número de columnas:", df.shape[1])

# Revisar cuántas muestras por clase tenemos
print("\nDistribución de clases:")
print(df["species"].value_counts())

Número de filas: 150
Número de columnas: 6

Distribución de clases:
species
setosa        50
versicolor    50
virginica     50
Name: count, dtype: int64


- Vemos que tenemos un total de 150 filas de datos:
  + Si usamos **70% train - 30% test**, serían 105 para entrenar y 45 para probar.
  + Si usamos** 80% train - 20% test**, serían 105 para entrenar y 30 para probar.

- Ambos son razonables, no obstante en datasets pequeños se prefiere hacer **80/20** para no quedarnos cortos de daos de entrenamiento.


In [8]:
# Separamos con sklearn

# Variables independientes (x) y dependiente (y)
X = df[iris.feature_names]
y = df["species"]

# Dividimos en train y test (80%/20%)
X_train, X_test, y_train, y_test = sk.model_selection.train_test_split(
    X, y, test_size= 0.2, random_state= 42, stratify=y
)

print("Tamaño entrenamiento:", X_train.shape[0])
print("Tamaño prueba", y_train.shape[0])

Tamaño entrenamiento: 120
Tamaño prueba 120


### **CREACIÓN Y AJUSTE DEL MODELO (ÁRBOLES DE DECISIÓN)**